File: app/src/main/java/com/cittus/isv/controller/MainImage.java
package com.cittus.isv.controller;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.*;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonArrayRequest;
import com.cittus.isv.R;
import com.cittus.isv.complements.slider.CustomVolleyRequest;
import com.cittus.isv.complements.slider.ViewPagerAdapter;
import com.cittus.isv.model.SliderUtils;
import com.cittus.isv.view.MainActivity;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class MainImage extends AppCompatActivity {

    // Variables Globals
    String request_url;
    Spinner spinner;
    Boolean sliderOn = false;

    // Variables locals - IMG
    ViewPager viewPager;
    LinearLayout sliderDotspanel;
    private int dotscount;
    private ImageView[] dots;
    private String imgURL;
    RequestQueue rq;
    List<SliderUtils> sliderImg;
    ViewPagerAdapter viewPagerAdapter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_image);
        // Init Process
        intiProcess();
    }

    private void intiProcess() {
        // Init Variables
        initVariables();

        // Get Elements Base (Consult and Set)
        getValuesMain();

        // Make Slider
        getRequest();

        // Set Process to Click
        processClick();

        // BTN Save
        saveElements();
    }

    private void initVariables() {

        // Init Elements Base (Consult and Set)
        sliderImg = new ArrayList<>();
        rq = CustomVolleyRequest.getInstance(this).getRequestQueue();

        // Init Elements - View
        viewPager = (ViewPager) findViewById(R.id.viewPager);
        sliderDotspanel = (LinearLayout) findViewById(R.id.SliderDots);

        // Selection of the spinner
        spinner = (Spinner) findViewById(R.id.spin_code_images);
    }

    private void getValuesMain() {
        // Intent Variable
        Intent intent = getIntent();

        // Set Template New Activity
        TextView title = findViewById(R.id.lbl_title_images);
        title.setText(intent.getStringExtra("title"));

        // Url Main Base - Imgs
        request_url = intent.getStringExtra("url_img");

        // Active Select
        findViewById(R.id.codeMain).setVisibility(intent.getBooleanExtra("code", true) ? View.VISIBLE : View.GONE);
        //findViewById(R.id.descriptionMain).setVisibility(intent.getBooleanExtra("description", true) ? View.VISIBLE : View.GONE);

        // Url Main Base - Imgs
        request_url = intent.getStringExtra("url_img");
    }

    private void processClick() {
        // Call the window of Main Image
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            // Set selected Action
            @Override
            public void onPageSelected(int position) {
                sliderOn = true;
                setElementByPosition(position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
        // Spinner Actions

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> arg0, View arg1, int position, long id) {
                // TODO Auto-generated method stub

                String ss = spinner.getSelectedItem().toString();
                for (int i = 0; i < arrayCodes.length; i++) {
                    if (arrayCodes[i] == ss) {
                            /*
                            int positionTemp = (i);
                            String imageUrl = arrayImages[positionTemp];
                            sliderImg = new ArrayList<>();
                            SliderUtils sliderUtils = new SliderUtils();
                            sliderUtils.setSliderImageUrl(imageUrl);
                            sliderImg.add(sliderUtils);
                            setSlider();*/
                    }
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> arg0) {
                // TODO Auto-generated method stub
            }
        });
    }

    String arrayCodes[];
    String arrayImages[];

    public void getRequest() {
        // Init Request JSOn
        JsonArrayRequest jsonArrayRequest = new JsonArrayRequest(Request.Method.GET, request_url, null, new Response.Listener<JSONArray>() {
            @Override
            public void onResponse(JSONArray response) {
                // Init Variables Base
                arrayCodes = new String[response.length()];
                arrayImages = new String[response.length()];
                for (int i = 0; i < response.length(); i++) {
                    // Make the Object Base
                    SliderUtils sliderUtils = new SliderUtils();
                    try {
                        // Get the object JSPM
                        JSONObject jsonObject = response.getJSONObject(i);
                        // Get And Set Data from te JSON
                        arrayCodes[i] = jsonObject.getString("code");
                        arrayImages[i] = jsonObject.getString("image_url");
                        imgURL = jsonObject.getString("image_url");
                        // Set Imagen parameter
                        sliderUtils.setSliderImageUrl(imgURL);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                    // Add Object to the Array
                    sliderImg.add(sliderUtils);
                }

                //setSpinner(arrayCodes, arrayImages);
                // Set Elements Get To the Slider
                setSlider();
                // Set Elements Get to the Spinner
                setSpinner();
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {

            }
        });
        // Make Request
        CustomVolleyRequest.getInstance(this).addToRequestQueue(jsonArrayRequest);
    }

    private void setSlider() {
        // Init Elements Base - Adapters
        viewPagerAdapter = new ViewPagerAdapter(sliderImg, MainImage.this, spinner);
        viewPager.setAdapter(viewPagerAdapter);
        // Get Count
        dotscount = viewPagerAdapter.getCount();
        // Check Elements
        if (dotscount > 0) {
            // Make Dots by Count
            dots = new ImageView[dotscount];
            for (int i = 0; i < dotscount; i++) {
                // Set location and move to do it
                dots[i] = new ImageView(MainImage.this);
                dots[i].setImageDrawable(ContextCompat.getDrawable(getApplicationContext(), R.drawable.non_active_dot));
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
                params.setMargins(8, 0, 8, 0);
                sliderDotspanel.addView(dots[i], params);
            }
            // Set image at select the dot
            dots[0].setImageDrawable(ContextCompat.getDrawable(getApplicationContext(), R.drawable.active_dot));
        }
    }

    private void setSpinner() {
        // Application of the Array to the Spinner
        ArrayAdapter<String> spinnerArrayAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, arrayCodes);
        spinnerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item); // The drop down view
        spinner.setAdapter(spinnerArrayAdapter);
    }

    private void setElementByPosition(int position) {
        // Move another Elements (DOTS)
        for (int i = 0; i < dotscount; i++) {
            dots[i].setImageDrawable(ContextCompat.getDrawable(getApplicationContext(), R.drawable.non_active_dot));
        }
        // Set Image
        dots[position].setImageDrawable(ContextCompat.getDrawable(getApplicationContext(), R.drawable.active_dot));
        // Set element in Spinner
        if (position >= 0) {
            spinner.setSelection(position);
        }
    }

    private void saveElements() {
        findViewById(R.id.btn_save_image).setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                ArrayList elementsBase = new ArrayList<String>();
                if (spinner != null) {
                    String ss = spinner.getSelectedItem().toString();

                    for (int i = 0; i < arrayCodes.length; i++) {
                        if (arrayCodes[i] == ss) {
                            elementsBase.add("'Number':'" + i + "'");
                            elementsBase.add("'Code':'" + arrayCodes[i] + "'");
                            elementsBase.add("'ImgSelected':'" + arrayImages[i] + "'");
                        }
                    }


                    // Array
                    Intent intent = getIntent();
                    intent.putExtra("getData", elementsBase);
                    setResult(5, intent);
                    finish();
                }
                return true;
            }
        });


    }
}

File: app/src/main/java/com/cittus/isv/controller/MainImage.kt
package com.cittus.isv.controller

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.ViewPager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView
import androidx.navigation.Navigation
import com.cittus.isv.R
import com.cittus.isv.model.*

class MainImage : Fragment() {

    // Main Variables
    private lateinit var viewMain: View
    // Make Bundle
    val bundle = Bundle()
    var login = 0
    private var municipalities: Municipalities? = null
    private var geolocationCardinalImages: ArrayList<GeolocationCardinalImages>? = null
    var signalArrayList = ArrayList<CittusISV>()
    private var size = 0

    // Variables locals - IMG
    private var imagenSelect: CittusImage? = null
    var mainImages = MainImages()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        viewMain = inflater.inflate(R.layout.activity_main_image, container, false)
        return viewMain
    }


    // TODO: Get Data - Municipalities
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val someDataClass: CittusListSignal? = arguments?.getParcelable("CittusDB")
        someDataClass?.let {
            login = it.login
            municipalities = it.municipality
            geolocationCardinalImages = it.geolocationCardinalImages
            signalArrayList = it.signal!!
        }

        // Get data Image
        imagenSelect = arguments?.getParcelable("CittusImage")

        if (login === 1) {
            size = signalArrayList.size - 1
            // Init Process
            initProcess()
        }
    }


    private fun initProcess() {

        var viewPager = viewMain!!.findViewById<ViewPager>(R.id.viewPager)
        var SliderDots = viewMain!!.findViewById<LinearLayout>(R.id.SliderDots)
        var spinCode = viewMain!!.findViewById<Spinner>(R.id.spin_code_images)

        // Start Process
        mainImages.getValuesMain(imagenSelect, context, viewPager, SliderDots, spinCode)

        // Active Select
        viewMain!!.findViewById<LinearLayout>(R.id.codeMain).visibility =
            View.VISIBLE//findViewById(R.id.codeMain).setVisibility(intent.getBooleanExtra("code", true) ? View.VISIBLE : View.GONE);

        // Set Template New Activity
        viewMain!!.findViewById<TextView>(R.id.lbl_title_images).text = imagenSelect!!.title

        saveElements()

    }


    internal var elementsBase = ArrayList<String>()

    private fun saveElements() {
        viewMain!!.findViewById<Button>(R.id.btn_save_image).setOnClickListener {
            // TODO: Get Data Images
            // 0 -> Number
            // 1 -> Code
            // 2 -> Img Select
            if (mainImages.spinner != null) {
                val ss = mainImages.spinner.selectedItem.toString()
                var arrayCodes = mainImages.arrayCodes
                var arrayImages = mainImages.arrayImages
                for (i in arrayCodes.indices) {
                    if (arrayCodes[i] === ss) {
                        elementsBase.add(0, i.toString())
                        elementsBase.add(1, arrayCodes[i])
                        elementsBase.add(2, arrayImages[i])
                    }
                }
                // Array Add Images
                var imagesByCode: ImagenSignalCode = ImagenSignalCode()
                imagesByCode.idImagen = elementsBase.get(0).toInt()
                imagesByCode.codeImagen = elementsBase.get(1)
                imagesByCode.pathImagen = elementsBase.get(2)

                signalArrayList[size].imagesByCode = imagesByCode

                // Make Object Main
                var cittusDB: CittusListSignal =
                    CittusListSignal(login, municipalities, signalArrayList, geolocationCardinalImages)
                // Set and Send Data Main
                Log.e("Data-Image", cittusDB.toString())
                bundle.putParcelable("CittusDB", cittusDB)
                Navigation.findNavController(viewMain!!).navigate(R.id.addressLocationOnTheWayActivity, bundle)
            }
        }
    }
}